docs/rfcs/20220207_connections_per_ip.md

```diff
@@ -0,0 +1,160 @@
+- Feature Name: Client conection limits
+- Status: in-progress
+- Start Date: 2022-02-07
+- Authors: John Spray
+- Issue: 
+
+# Executive Summary
+
+Add per-client state to track and limit the number of connections open.
+
+Similar to KIP-308, adapted to Redpanda's thread per core model.
+
+## What is being proposed
+
+New configuration properties:
+- kafka_max_connections_per_ip (optional integer, default unset)
+- kafka_max_connections_per_ip_overrides (string, default "")
+
+When a connection is accepted which exceeds one of the configured
+bounds, it is dropped before reading or writing anything to the socket.
```

Comment:
It would be super awesome if we could use eBPF to drop the connections at line rate. This is an example of how it could be done: https://github.com/cloudflare/cloudflare-blog/blob/master/2018-07-dropping-packets/bpf-drop.c. https://blog.cloudflare.com/how-to-drop-10-million-packets/ has more details.

Reply:
If it is too involved or feels out of place, we can also do it for you later on outside of Redpanda.

Reply:
I've added an alternatives-considered section on this -- I think using kernel helpers for this right now is probably too much (we are mainly guarding against misbehaving apps rather than hostile DDoS), and we need the userspace version anyway for any environments where we don't have access to the kernel bits (old kernels, locked down environments).